### DISCUSSION TITLE
Warp-style terminal features

### DISCUSSION THREAD

**Discussion by austincondiff** [migueldeicaza/SwiftTerm#283] [Ideas]
I am really liking the newly-released [Warp](https://www.warp.dev/). I wonder if it would be possible to bring some of it's most simple features to this project as optional configurations. These might include:

- Terminal blocks (input+output) separated by a line
- Terminal input sticky header
- Most recent prompt stuck to the bottom similar to Messages (reverse scroll direction)
- Current input is sticky footer
- Smooth scroll (instead of line by line) #260

**Comment by migueldeicaza** [comment]
I need to research warp to understand what exactly these things do as it is not clear from the description.

There are elements like the blocks that I saw on their videos that look interesting - but also are fairly different from what a terminal emulator does, and I believe this behavior is sort of a pre-processor.  It relies on a few pieces:

1. You need to be able to track the state of the input - and this is typically done with assorted scripts that your shell runs on your behalf.   In this case, you just want to know when to trigger an alternative input system.

2. The alternative input system: this is whatever you want it to be.  It could be a separate window for instance or a neatly embedded one that appears to be part of the terminal.  And when you are done, you send it, and you get out of the mode "I am editing on behalf of the shell" until you are notified by (1) again.

Midnight commander has been using a scheme similar to this since 1993 or so.  It works, but comes with some brittle edges that tend to piss of people every once in a while.

**Comment by migueldeicaza** [comment]
Ok I watched the video for some of these features.   Some of these are hard to retrofit into the terminal, but not impossible.

Let me think about how I would implement this and share some directional thoughts.

**Comment by migueldeicaza** [comment]
I had a chance to play with Warp now, I am plotting a plan.
